gcloud-java-dns/src/test/java/com/google/gcloud/dns/ITDnsTest.java

```diff
@@ -0,0 +1,1268 @@
+    assertEquals(ZONE1.name(), created.name()); // always returned
+    assertNull(created.creationTimeMillis());
+    assertNull(created.dnsName());
+    assertNull(created.description());
+    assertFalse(created.nameServers().isEmpty());
+    assertNull(created.nameServerSet());
+    assertNull(created.id());
+    created = DNS.getZone(ZONE1.name(), Dns.ZoneOption.fields(Dns.ZoneField.ZONE_ID));
+    assertEquals(ZONE1.name(), created.name()); // always returned
+    assertNull(created.creationTimeMillis());
+    assertNull(created.dnsName());
+    assertNull(created.description());
+    assertNotNull(created.nameServers());
+    assertTrue(created.nameServers().isEmpty()); // never returns null
+    assertNotNull(created.id());
+    // combination of multiple things
+    created = DNS.getZone(ZONE1.name(), Dns.ZoneOption.fields(Dns.ZoneField.ZONE_ID,
+        Dns.ZoneField.NAME_SERVERS, Dns.ZoneField.NAME_SERVER_SET, Dns.ZoneField.DESCRIPTION));
+    assertEquals(ZONE1.name(), created.name()); // always returned
+    assertNull(created.creationTimeMillis());
+    assertNull(created.dnsName());
+    assertEquals(ZONE1.description(), created.description());
+    assertFalse(created.nameServers().isEmpty());
+    assertNull(created.nameServerSet()); // we did not set it
+    assertNotNull(created.id());
+  }
+
+  @Test
+  public void testListZones() {
+    List<Zone> zones = filter(DNS.listZones().iterateAll());
+    assertEquals(0, zones.size());
+    // some zones exists
+    Zone created = DNS.create(ZONE1);
+    zones = filter(DNS.listZones().iterateAll());
+    assertEquals(created, zones.get(0));
+    assertEquals(1, zones.size());
+    created = DNS.create(ZONE_EMPTY_DESCRIPTION);
+    zones = filter(DNS.listZones().iterateAll());
+    assertEquals(2, zones.size());
+    assertTrue(zones.contains(created));
+    // error in options
+    try {
+      DNS.listZones(Dns.ZoneListOption.pageSize(0));
+    } catch (DnsException ex) {
+      // expected
+      assertEquals(400, ex.code());
+      // todo(mderka) test not-retryable
+    }
+    try {
+      DNS.listZones(Dns.ZoneListOption.pageSize(-1));
+    } catch (DnsException ex) {
+      // expected
+      assertEquals(400, ex.code());
+      // todo(mderka) test not-retryable
+    }
+    // ok size
+    zones = filter(DNS.listZones(Dns.ZoneListOption.pageSize(1000)).iterateAll());
+    assertEquals(2, zones.size()); // we still have only 2 zones
+    // dns name problems
+    try {
+      DNS.listZones(Dns.ZoneListOption.dnsName("aaaaa"));
+    } catch (DnsException ex) {
+      // expected
+      assertEquals(400, ex.code());
+      // todo(mderka) test not-retryable
+    }
+    // ok name
+    zones = filter(DNS.listZones(Dns.ZoneListOption.dnsName(ZONE1.dnsName())).iterateAll());
+    assertEquals(1, zones.size());
+    // field options
+    zones = ImmutableList.copyOf(DNS.listZones(Dns.ZoneListOption.dnsName(ZONE1.dnsName()),
```

Comment:
Here and in the following: is the `ImmutableList.copyOf` step necessary? Couldn't this be done using the iterator?